### ISSUE TITLE
(Bounty) Allow usage of own Firefox Sync token server

### ISSUE THREAD

**Issue by schrorg** [mozilla-mobile/firefox-ios#1276]
We use an own Firefox Sync server like described in https://docs.services.mozilla.com/howtos/run-sync-1.5.html. But there is no way to configure the Firefox Sync token server URL to point to our own sync server other than modifying the source code in `Account/FirefoxAccountConfiguration.swift`, line 111.

Can we add an option for that? Should I do a PR?

**Comment by egogo-nl** [comment]
I would also like to see this feature added.

**Comment by acaranta** [comment]
bump .... That would really be a enhancement ....

**Comment by deruli79** [comment]
Agree. This is really the main feature for me to leave Safari behind...

**Comment by st3fan** [comment]
@schrorg We would love to see a PR for this. Do you want to work on this?

**Comment by dprandzioch** [comment]
Has there been any progress on this topic? I'd really love to see this feature as FF would be the only browser to support self-hosted sync across all necessary devices.

**Comment by schrorg** [comment]
Unfortunately I will not have the time for it during the next weeks (summer vacation, lots of other work to do...).
I will definitely keep this in mind because I still need this feature personally.

**Comment by dprandzioch** [comment]
I'm not that experienced in iOS development although I did some work in own apps already. If there's something I can help you with, please let me know.

**Comment by mehdicopter** [comment]
Any improvements on this feature ?
I'm also interested to point to my self-hosted ffsync server.
Thanks.

**Comment by egogo-nl** [comment]
Are there any updates on this? I would really like to use this feature.

**Comment by kwitaszczyk** [comment]
+1

**Comment by iron77** [comment]
I would love to see this feature being implemented as well. Shall we put a bounty on it? Any specific service you guys would recommend for this?

**Comment by mrjovanovic** [comment]
+1 Would chip in on the bounty as well.

**Comment by iron77** [comment]
OK, I just started it out at Bountysource. Feel free to contribute! :)

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=28349461)](https://www.bountysource.com/issues/28349461-allow-usage-of-own-firefox-sync-token-server?utm_source=28349461&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by dsabanin** [comment]
+1 I joined in.

**Comment by iron77** [comment]
@schrorg Would you consider renaming the issue topic to include "(Bounty)" in the front, so it would be more noticeable for anybody who just looks at the list of issues?

@mrjovanovic Now you have a chance :).

@ All other guys, if it's something you want to see being done soon, I'd suggest to drop few bucks on it as well :). 

Also, I wonder if there's a good place to advertise the issue once the bounty gets big enough? Maybe Mozilla iOS IRC?

**Comment by m8mble** [comment]
I've filed bug [1344536](https://bugzilla.mozilla.org/show_bug.cgi?id=1344536) for this. Starting to work on it...

**Comment by hb2708** [comment]
Hello All.
@schrorg @iron77 
To Cut the story short we all want the option to change the 

`Sync15Configuration  -> tokenServerEndpointURL`  URL From the UI front.

Please Correct me if i am wrong.

Would Love to start working on the same. If any one have not yet started.

**Comment by cdubz** [comment]
Note that [1344536](https://bugzilla.mozilla.org/show_bug.cgi?id=1344536) has been marked as a duplicate of [1204411](https://bugzilla.mozilla.org/show_bug.cgi?id=1204411) and closed. @st3fan, please remove the `needs bug` label.

**Comment by iron77** [comment]
@hb2708 thanks for your interest.

I might not be fully aware of how this iOS app works under the hood, but from the user perspective, I can tell you which configuration options we need there to be present.

Bringing desktop Firefox as an example, to run own Firefox Sync server, you go to `about:config` and set up some options pointing to where your server is. As per https://mozilla-services.readthedocs.io/en/latest/howtos/run-fxa.html and some of my testing, the options are:

- identity.fxaccounts.auth.uri
- identity.fxaccounts.remote.force_auth.uri
- identity.fxaccounts.remote.oauth.uri
- identity.fxaccounts.remote.profile.uri
- identity.fxaccounts.remote.signin.uri
- identity.fxaccounts.remote.signup.uri
- identity.fxaccounts.remote.webchannel.uri
- identity.fxaccounts.settings.uri
- identity.sync.tokenserver.uri

I'd expect iOS app (both iPhone and iPad) to also give you some way, somewhere in UI, to set these options to custom values. 

For the record, it was about a year since I last digged in that documentation, so in case I missed any option which might also be required these days, I'd appreciate the comment.

**Comment by hb2708** [comment]
@m8mble  @schrorg  @iron77  Thanks but I am not aware of it on desktop.
But i can surely make it possible on iPhone and iPad apps.

If any can explain is some easy manner i would be grateful.

@cdubz So it means it's implemented or yet to be implemented?

**Comment by cdubz** [comment]
Still yet to be implemented and doesn't appear to have any traction in Bugzilla. I was just noting the updated bug number there.

**Comment by iron77** [comment]
@hb2708 I can try to explain it as easy as possible. Just so I better understand where to start, can you tell me:
1. Did you ever used Firefox Sync:
   - Mozilla hosted? (default)
   - Self-hosted?
2. If none of the above, do you have a basic understanding of what Firefox Sync is?

Sorry if questions above sound a bit stupid, I want to be fully sure we're both on the same page :)

**Comment by cwm88** [comment]
Hi,
the only setting which needs to be changed to be editable is the analogous option to **identity.sync.tokenserver.uri**. In all other OSes (MacOS,WIN,Linux,Android) it was seemingly implemented at a easy way, only in iOS it seems not to be hassle-free. FF at this OS has obviously not as much in common with the other Implementations. A bing thank you in advance to the people who are in charge of this!

**Comment by rgschi-tgm** [comment]
Giving this another bump since I just upgraded to iOS10 from 7 to get Firefox on my iP. Now this missing function disappoints me a bit. 

It cannot be in any way too hard to code, seems like a couple of lines only.
The problem I see would be getting this code into a release. Are the maintainers willing to accept "foreign" code?

**Comment by joeyg** [comment]
Is this issue resolved now that https://github.com/mozilla-mobile/firefox-ios/pull/3088 has been merged or is additional functionality that needs to be added?

**Comment by iron77** [comment]
@joeyg I believe this is it! Looking forward to test it at my iDevices once it lands the App Store.

**Comment by st3fan** [comment]
Yes, I am going to close this issue now that https://github.com/mozilla-mobile/firefox-ios/pull/3088 has landed. I think it has only landed on master though. @vbudhram is #3088 something we want to uplift to 9.0?
